zebra-state/src/service.rs

```diff
@@ -679,3 +690,43 @@ impl Service<Request> for StateService {
 pub fn init(config: Config, network: Network) -> BoxService<Request, Response, BoxError> {
     BoxService::new(StateService::new(config, network))
 }
+
+/// Check if zebra is following a legacy chain and return an error if so.
+pub fn legacy_chain_check<I>(
+    tip_height: block::Height,
+    ancestors: I,
+    network: Network,
+) -> Result<(), BoxError>
+where
+    I: Iterator<Item = Arc<Block>>,
+{
+    const MAX_BLOCKS_TO_CHECK: usize = 100;
+
+    // Only do the check if we have an Nu5 activation height.
+    if let Some(nu5_height) = NetworkUpgrade::Nu5.activation_height(network) {
+        // And only do the check if we are above Nu5 activation.
+        if tip_height >= nu5_height {
```

Comment:
1. Bug: Do this check on every block height, not just the first block height

2. Explain design: Why is this check is important? Why does it need to be done for every block?

3. Confusing code: Let's call it `nu5_activation_height`, because there are lots of NU5 blocks

```suggestion
        // Stop checking if the chain reaches Canopy. We won't find any more V5 transactions,
        // so the rest of our checks are useless.
        //
        // If the cached tip is close to NU5 activation, but there aren't any V5 transactions in the
        // chain yet, we could reach MAX_BLOCKS_TO_CHECK in Canopy, and incorrectly return an error.
        if block.coinbase_height().expect("valid blocks have coinbase heights") < nu5_activation_height {
            return Ok(());
        }
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/2366/commits/8bb38d5ff2fa83dbabd5b1b7158148be0f475f1c